plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GradleIntegrationHarness.java

```diff
@@ -59,6 +60,10 @@
 		}
 	}
 
+	public static <T> Provider<T> providerOf(T value) {
+		return org.gradle.api.internal.provider.Providers.of(value);
```

Comment:
I remember having used the following in a Gradle build script recently:

```kotlin
val someList = provider {
    listOf(1, 2, 3)
}
```

So I was wondering if this `provider` function is available outside of the Gradle DSL. If so, it would mean we wouldn't need to use an internal API here.

Reply:
This is a good question, and the answer is super surprising -`interface org.gradle.api.provider.Provider` is public, and has [quite a few methods](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Provider.html). But there are literally *no* implementations (abstract or not) in an api package, only internal (that I could find at least).

If you take the time to implement it yourself (which I did), then you actually get a ClassCastException! Because somewhere in Gradle's guts, they cast it to `org.gradle.internal.ProviderImpl` (I forget the exact name), which means you can't implement it yourself.

We have a code style rule that barfs at any `.internal.` imports, which is why I spelled-out the package inline like I did. I'm still a huge gradle fan, but I was pretty bummed when I found out that you can't actually implement `Provider` 😞

Reply:
That _is_ surprising! Maybe the Gradle developers wanted something like sealed classes for `Provider` but couldn't get it for the version of Java they're constrained to.

I would've also guessed that `Provider` was a single-method interface, where all the other methods were `default`. But maybe it's because the minimum Java version that Gradle runs on isn't Java 8 yet.

Having done some research just now, I see that `provider { }` is indeed only available in the DSL, specifically on `Project.getProviders()`, `Settings.getProviders()`, and through `@Inject`ing a `ProviderFactory` into a plugin or extension. I suppose if we weren't constrained by the configuration cache we'd be able to use one of them, but alas.

Thanks for explaining!